### PR TITLE
Allow specification of openshift_additional_registry_credentials

### PR DESCRIPTION
**Pull request by sdodson**
- Corrects docker_creds test_login type
- Adds tls_verify option to docker_creds
- Adds openshift_additional_registry_credentials with the ability to specify host, user, password, test_login, image_name, and tls_verify parameters

Should be in the form of
openshift_additional_registry_credentials=[{'host':'registry.example.com','user':'bob','password':'pass1','test_login':'False'},{'host':'registry2.example.com','password':'pass2','tls_verify':'False','image_name':'openshift3/ose-pod'}]

If the user is not specified it will default to 'openshift', this can be used in situations where a token is all that's required for a given host but skopeo requires a username.

Care should be taken to ensure that you do not specify conficting
credentials with those set via oreg_url, oreg_auth_password,
oreg_auth_user.

In the future we should clean this up to de-dupe these
two sets of credentials and eliminate the duplicated tasks.

### CODE REVIEW

**Comment by sdodson** [comment]
Untested, just getting this up to discuss UX.

File: roles/nuage_master/tasks/serviceaccount.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@
     namespace: default
     resource_name: "{{ item.resource_name }}"
     resource_kind: "{{ item.resource_kind }}"
-    user: "{{ item.user }}"
+    user: "{{ item.user | default('openshift') }}"
```
Comment (michaelgugino) [inline roles/nuage_master/tasks/serviceaccount.yml]: Is this change intentional?

File: roles/openshift_node/tasks/registry_auth.yml [hunk 2]
```diff
@@ -30,6 +30,25 @@
...
+  - openshift_additional_registry_credentials is defined
+  register: crt_oreg_auth_credentials_create
+  retries: 3
+  delay: 5
+  until: crt_oreg_auth_credentials_create is succeeded
+  with_items:
+    - "{{ openshift_additional_registry_credentials | default({}) }}"
+
 # Container images may need the registry credentials
 - name: Setup ro mount of /root/.docker for containerized hosts
```
Comment (michaelgugino) [inline roles/openshift_node/tasks/registry_auth.yml]: Need to modify the booleans here to account for this new task.

File: roles/openshift_control_plane/tasks/registry_auth.yml [hunk 3]
```diff
@@ -19,3 +19,22 @@
...
   until: master_oreg_auth_credentials_create is succeeded
+
+- name: Create credentials for any additional registries
+  docker_creds:
+    path: "{{ docker_cli_auth_config_path }}"
+    registry: "{{ item.host }}"
+    username: "{{ item.user | default('openshift') }}"
+    password: "{{ item.password }}"
+    # Test that we can actually connect with provided info
+    test_login: "{{ oreg_test_login | default(True) }}"
```
Comment (michaelgugino) [inline roles/openshift_control_plane/tasks/registry_auth.yml]: Might want fine-grained control for test_login in this dictionary because skopeo will probably fail on self-signed cert registries; not sure if skopeo respects docker insecure registries, need to confirm.

File: roles/container_runtime/tasks/registry_auth.yml [hunk 4]
```diff
@@ -18,3 +18,22 @@
...
+    registry: "{{ item.host }}"
+    username: "{{ item.user | default('openshift') }}"
+    password: "{{ item.password }}"
+    # Test that we can actually connect with provided info
+    test_login: "{{ oreg_test_login | default(True) }}"
+    proxy_vars: "{{ l_docker_creds_proxy_vars }}"
+    image_name: "{{ l_docker_creds_image_name }}"
+  when:
+  - openshift_additional_registry_credentials is defined
+  register: crt_oreg_auth_credentials_create
```
Comment (michaelgugino) [inline roles/container_runtime/tasks/registry_auth.yml]: Reusing registered variable names, while not problematic at this case (AFAICT), can create problems.  Don't do that.

File: roles/nuage_master/tasks/serviceaccount.yml (hunk 1 above)
Comment (sdodson) [reply roles/nuage_master/tasks/serviceaccount.yml]: No it wasn't, thanks for checking.

File: roles/openshift_control_plane/tasks/registry_auth.yml (hunk 3 above)
Comment (sdodson) [reply roles/openshift_control_plane/tasks/registry_auth.yml]: I'll add both test_login and tls_verify options. Defaulting both to true.

File: roles/lib_utils/library/docker_creds.py [hunk 5]
```diff
@@ -131,14 +131,15 @@ def load_config_file(module, dest):
         return {}
 
 
-def gen_skopeo_cmd(registry, username, password, proxy_vars, image_name):
+def gen_skopeo_cmd(registry, username, password, proxy_vars, image_name, tls_verify):
```
Comment (michaelgugino) [inline roles/lib_utils/library/docker_creds.py]: You need to pass this param to the function call as well.

File: roles/lib_utils/library/docker_creds.py [hunk 6]
```diff
@@ -131,14 +131,15 @@ def load_config_file(module, dest):
...
     '''Generate skopeo command to run'''
     skopeo_temp = ("{proxy_vars} timeout 10 skopeo inspect"
                    " {creds} docker://{registry}/{image_name}")
     # this will quote the entire creds argument to account for special chars.
     creds = pipes.quote('--creds={}:{}'.format(username, password))
     skopeo_args = {'proxy_vars': proxy_vars, 'creds': creds,
-                   'registry': registry, 'image_name': image_name}
+                   'registry': registry, 'image_name': image_name,
+                   'tls_verify': tls_verify}
```
Comment (michaelgugino) [inline roles/lib_utils/library/docker_creds.py]: This will result in just 'True' or 'False' being passed to the command, need to format the command similar to creds.

File: roles/openshift_examples/tasks/main.yml [hunk 7]
```diff
@@ -49,18 +49,26 @@
...
   when:
     - openshift_examples_load_rhel | bool
     - oreg_auth_password is defined
-  with_items:
-    - "{{ rhel_image_streams }}"
   register: oex_imagestream_import_secret
   failed_when: "'already exists' not in oex_imagestream_import_secret.stderr and oex_imagestream_import_secret.rc != 0"
   changed_when: false
 
+- name: Create imagestream import secrets for any additional registries
```
Comment (sdodson) [inline roles/openshift_examples/tasks/main.yml]: @bparees I seem to recall you suggested we only provision credentials for openshift_examples_registryurl into the openshift namespace or am I mis-remembering? Should the openshift namespace get the additional credentials too?

File: roles/openshift_examples/tasks/main.yml [hunk 8]
```diff
@@ -49,18 +49,26 @@
...
+  command: >
+      {{ openshift_client_binary }} create secret docker-registry imagestreamsecret --docker-server={{ item.host }} --docker-username={{ item.user }} --docker-email=openshift@openshift.com --docker-password={{ item.password }} --config={{ openshift.common.config_base }}/master/admin.kubeconfig -n openshift
```
Comment (michaelgugino) [inline roles/openshift_examples/tasks/main.yml]: Break up this line for readability.  No sense in using '>' if we're not going to utilize multiple lines.

File: roles/openshift_node/tasks/registry_auth.yml [hunk 9]
```diff
@@ -30,13 +30,34 @@
...
   set_fact:
     l_bind_docker_reg_auth: True
   when:
     - openshift_is_atomic | bool
-    - oreg_auth_user is defined
+    - oreg_auth_user is defined or openshift_additional_registry_credentials is defined
```
Comment (michaelgugino) [inline roles/openshift_node/tasks/registry_auth.yml]: I think this would be better as
```yaml
- openshift_is_atomic | bool
- oreg_auth_user is defined or openshift_additional_registry_credentials is defined or node_oreg_auth_credentials_stat.stat.exists
```

Logic:  If we defined oreg_auth_user or openshift_additional_registry_credentials, we know for certain we will be updating/creating the .docker/config.json file.  If node_oreg_auth_credentials_stat.stat.exists, then it is already on disk (someone created that outside of openshift-ansible, etc), we don't need to check for changed because changed will be covered by defining the auth variables.

Comment (sdodson) [reply roles/openshift_node/tasks/registry_auth.yml]: ok, updated

File: roles/openshift_examples/tasks/main.yml (hunk 8 above)
Comment (sdodson) [reply roles/openshift_examples/tasks/main.yml]: ok

File: roles/container_runtime/tasks/registry_auth.yml [hunk 10]
```diff
@@ -18,3 +18,23 @@
...
+    image_name: "{{ item.image_name | default('openshift3/ose-pod') }}"
+    tls_verify: "{{ item.tls_verify | default(omit) }}"
+  when:
+  - openshift_additional_registry_credentials is defined
+  register: crt_addl_credentials_create
+  retries: 3
+  delay: 5
+  until: crt_addl_credentials_create is succeeded
+  with_items:
+    "{{ openshift_additional_registry_credentials }}"
```
Comment (michaelgugino) [inline roles/container_runtime/tasks/registry_auth.yml]: I'm not 100% sure, but I think this should be in-line with with_items.
